reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala

```diff
@@ -67,57 +71,75 @@ private[reactivestreams] final class StreamSubscription[F[_], A](
         go(in).stream
       }
 
-    val s =
-      stream
-        .through(subscriptionPipe)
-        .interruptWhen(cancelled)
-        .evalMap(x => F.delay(sub.onNext(x)))
-        .handleErrorWith(e => Stream.eval(onError(e)))
-        .onFinalize(cancelled.get.ifM(ifTrue = F.unit, ifFalse = onComplete))
-        .compile
-        .drain
-
-    startDispatcher.unsafeRunAndForget(s)
+    stream
+      .through(subscriptionPipe)
+      .interruptWhen(cancelled)
+      .evalMap(x => F.delay(sub.onNext(x)))
+      .handleErrorWith(e => Stream.eval(onError(e)))
+      .onFinalize(cancelled.get.ifM(ifTrue = F.unit, ifFalse = onComplete))
+      .compile
+      .drain
   }
 
   // According to the spec, it's acceptable for a concurrent cancel to not
   // be processed immediately, but if you have synchronous `cancel();
   // request()`, then the request _must_ be a no op. Fortunately,
-  // ordering is guaranteed by a sequential d
+  // ordering is guaranteed by a sequential dispatcher.
   // See https://github.com/zainab-ali/fs2-reactive-streams/issues/29
   // and https://github.com/zainab-ali/fs2-reactive-streams/issues/46
-  def cancel(): Unit =
-    requestDispatcher.unsafeRunAndForget(cancelled.set(true))
-
-  def request(n: Long): Unit = {
-    val request: F[Request] =
-      if (n == java.lang.Long.MAX_VALUE) (Infinite: Request).pure[F]
-      else if (n > 0) (Finite(n): Request).pure[F]
-      else F.raiseError(new IllegalArgumentException(s"3.9 - invalid number of elements [$n]"))
-
-    val prog = cancelled.get
-      .ifM(ifTrue = F.unit, ifFalse = request.flatMap(requests.offer).handleErrorWith(onError))
+  private def cancelMe: F[Unit] =
+    F.delay(completed.set(true)) >> cancelled.set(true)
+  override def cancel(): Unit =
+    if (!completed.get()) {
+      requestDispatcher.unsafeRunSync(cancelMe)
```

Comment:
@armanbilge Mostly to ensure future `cancel` and `request` calls to be NOOPs. Thus I ensure that `subscribe` waits for the `completed` and `cancelled` flags to be set.
And since those two operations are pretty fast, we are not imposing a hard block on our callers.

Reply:
> Mostly to ensure future `cancel` and `request` calls to be NOOPs. Thus I ensure that `subscribe` waits for the `completed` and `cancelled` flags to be set.

Thanks, but it's not obvious to me why guaranteeing this behavior this requires running this action synchronously. According to the comment the spec says it may run asynchronously.

---

> And since those two operations are pretty fast, we are not imposing a hard block on our callers.

`unsafeRunSync()` always forces a shift across threadpools and possibly creating a new thread for blocking. It's hardly a trivial operation.

Reply:
Is the `completed.get()` intended to guard against invalid invocations of the `Dispatcher`? If so, I think there is a race condition here with concurrent invocations of `cancel()` and `request()`.

Reply:
> Thanks, but it's not obvious to me why guaranteeing this behavior this requires running this action synchronously. According to the comment the spec says it may run asynchronously.

Yes, it may run asynchronously.
However, if a `Subscriber` does something like:

```scala
subscription.cancel()
subscription.request()
```

That `request` HAS to be a NOOP, and if we don't block the `cancel` to set `completed` before returning, then it may not be a NOOP.
I don't remember, but I think I had a failing test because of that.

-----

> unsafeRunSync() always forces a shift across threadpools and possibly creating a new thread for blocking.

I was under the impression that shouldn't be the case with a `sequential` `Dispatcher`, since AFAIK, those were optimized for this very use case, running something simple like `Queue.offer` synchronously.

-----

> Is the completed.get() intended to guard against invalid invocations of the Dispatcher?

Yes, although is a bit more general.
The idea is that `completed` is set when we reach what the spec calls a _"terminal state"_, after which all operations are NOOPs.

> If so, I think there is a race condition here with concurrent invocations of cancel() and request().

😿
That is probably the bug we are seeing, right?

Reply:
> However, if a `Subscriber` does something like:
> 
> ```scala
> subscription.cancel()
> subscription.request()
> ```
> 
> That `request` HAS to be a NOOP, and if we don't block the `cancel` to set `completed` before returning, then it may not be a NOOP.

This ordering will be preserved by the sequential dispatcher :) so we just need to check the condition in the submitted effect, and no-op it if appropriate.

---

> I was under the impression that shouldn't be the case with a `sequential` `Dispatcher`, since AFAIK, those were optimized for this very use case, running something simple like `Queue.offer` synchronously.

No, I'm afraid that's fundamental: you are asynchronously submitting an effect to the `IORuntime` from possibly outside the runtime, and blocking while you await its completion.

The optimization of a sequential dispatcher is that it runs all submitted effects on the same fiber, which is why it can only run one effect at a time and preserves order of (non-concurrent) submission.

The parallel dispatcher creates a new fiber for each submitted effect, which is overhead for simple operations as you point out.

---

> That is probably the bug we are seeing, right?

Possibly 😅 is there a situation when the `Dispatcher` is shutdown that we _shouldn't_ no-op? Seems like we could just ignore this exception.